docs/madr/decisions/081-datasource-api.md

```diff
@@ -0,0 +1,256 @@
+Let’s consider two examples: `MeshIdentity` and `MeshExternalService`.
+
+`MeshIdentity`
+When a user creates a `MeshIdentity`, they are expected to provide a certificate authority (CA) and a private key. These values are read by the control plane and used to sign workload certificates.
+This behavior is control-plane centric and secure by design.
+
+`MeshExternalService`
+In contrast, for `MeshExternalService`, it's unclear whether:
+* The control plane should resolve the DataSource (e.g., read `Secret`, `File`, or `EnvVar`) and pass it to Envoy via SDS, or
+* The generated Envoy config should directly reference local paths or environment variables (which must be pre-mounted or pre-set on the ZoneEgress proxy).
+
+This inconsistency makes the API harder to reason about, especially when using a shared structure.
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshIdentity
+metadata:
+  name: identity
+  namespace: kuma-system # only in system namespace
+  labels:
+    kuma.io/mesh: default
+spec:
+  ...
+  provider:
+    type: Bundled
+    bundled: # to extend in KM
+      ca:
+        certificate:
+          type: File
+          file:
+            path: "/etc/ssl/certs/my-cert.crt"
+        privateKey:
+          type: Secret
+          secretRef:
+            kind: Secret
+            name: kuma-private-key
+```
+
+In case of `MeshExternalServie` it doesn't seem to be clear if control-plane should read the DataSource or maybe it should be handled by the Envoy.
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshExternalService
+metadata:
+  name: mes-tcp-mtls
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+spec:
+  ...
+  tls:
+    enabled: true
+    verification:
+      serverName: tcpbin.com
+      clientCert:
+        type: File
+        file:
+          path: "/etc/ssl/certs/my-cert.crt"
+      clientKey:
+        type: Secret
+        secretRef:
+          kind: Secret
+          name: kuma-private-key
+```
+
+The API is not fully clear in this case. Specifically, it’s unclear whether the resource needs to be provided. Should it be:
+* Read by the control plane and delivered to `ZoneEgress` via `SDS`, or
+* Used to generate Envoy configuration where:
+  * EnvVar and File result in a cluster referencing the path or environment variable on the `ZoneEgress` (pod/VM),
+  * Secret is read by the control plane and passed via SDS.
+
+This results in a divergence in behavior across resources. For example:
+* In MeshIdentity, the control plane always reads the source and delivers it.
+* In MeshExternalService, it’s unclear whether the control plane or the data plane (`ZoneEgress`) is responsible.
+
+**How can we solve it?**
+
+To resolve the ambiguity around how and where a `DataSource` is accessed (i.e., control plane vs data plane), we could introduce a resource specific ResourceDescriptor parameter. This parameter would define whether the `DataSource` should be:
```

Comment:
would that parameter be somewhere on the data source? something similar to:

```
type: File
resolve: onCp | onDp 
file:
  path: "/etc/ssl/certs/my-cert.crt"
```

? we could then validate it on specific resource (MES / MID)